xmtp_db/src/sql_key_store/transactions.rs

```diff
@@ -0,0 +1,186 @@
+/// This connection _must only_ be created from starting a transaction
+pub struct MutableTransactionConnection<'a, C> {
+    // we cannot avoid interior mutability here
+    // because raw_query methods require &self, as do MlsStorage trait methods.
+    // Since we no longer have async transactions, once a transaction is started
+    // we can ensure it occurs all on one thread.
+    pub(crate) conn: RefCell<&'a mut C>,
+}
+
+impl<'a, C> MutableTransactionConnection<'a, C> {
+    pub fn new(conn: &'a mut C) -> Self {
+        Self {
+            conn: RefCell::new(conn),
+        }
+    }
+}
+
+impl<'a, C> ConnectionExt for MutableTransactionConnection<'a, C>
+where
+    C: diesel::Connection<Backend = Sqlite>
+        + diesel::connection::SimpleConnection
+        + LoadConnection
+        + MigrationConnection
+        + MigrationHarness<<C as diesel::Connection>::Backend>
+        + crate::MlsKeyStore
+        + Send,
+{
+    type Connection = C;
+
+    fn start_transaction(&self) -> Result<TransactionGuard, crate::ConnectionError> {
+        Err(crate::ConnectionError::Database(
+            diesel::result::Error::AlreadyInTransaction,
+        ))
+    }
+
+    fn raw_query_read<T, F>(&self, fun: F) -> Result<T, crate::ConnectionError>
+    where
+        F: FnOnce(&mut Self::Connection) -> Result<T, diesel::result::Error>,
+        Self: Sized,
+    {
+        let mut conn = self.conn.borrow_mut();
+        fun(&mut conn).map_err(crate::ConnectionError::from)
+    }
+
+    fn raw_query_write<T, F>(&self, fun: F) -> Result<T, crate::ConnectionError>
+    where
+        F: FnOnce(&mut Self::Connection) -> Result<T, diesel::result::Error>,
+        Self: Sized,
+    {
+        let mut conn = self.conn.borrow_mut();
+        fun(&mut conn).map_err(crate::ConnectionError::from)
+    }
+
+    fn is_in_transaction(&self) -> bool {
+        true
+    }
+
+    fn disconnect(&self) -> Result<(), crate::ConnectionError> {
+        // generally does not make sense
+        Ok(())
+    }
+
+    fn reconnect(&self) -> Result<(), crate::ConnectionError> {
+        // generally does not make sense
+        Ok(())
+    }
+}
+
+impl<C: ConnectionExt> XmtpMlsStorageProvider for SqlKeyStore<C> {
+    type Connection = C;
+
+    type DbQuery<'a>
+        = DbConnection<&'a C>
+    where
+        Self::Connection: 'a;
+
+    fn db<'a>(&'a self) -> Self::DbQuery<'a> {
+        DbConnection::new(&self.conn)
+    }
+
+    fn transaction<T, E, F>(&self, f: F) -> Result<T, E>
+    where
+        F: FnOnce(&mut <C as ConnectionExt>::Connection) -> Result<T, E>,
+        E: From<diesel::result::Error> + From<crate::ConnectionError> + std::error::Error,
+    {
+        let conn = &self.conn;
+
+        let _guard = self.conn.start_transaction(); // still needed so any reads use tx
+        // one call to raw_query_write = mutex only locked once for entire transaciton
+        let r = conn.raw_query_write(|c| Ok(c.transaction(|sqlite_c| f(sqlite_c))))?;
```

Comment:
Now that we only use one lock for a transaction (rather than locking three times), we can put all sqlite transactions into a RwLock, instead of a RwLock & Mutex